### ISSUE TITLE
AutoTokenizer silently corrupts deepseek-math-7b-rl in 5.x: encode strips all spaces, decode emits literal byte-level markers

### ISSUE THREAD

**Issue by johnmarktaylor91** [huggingface/transformers#48206]
### System Info

- `transformers` 5.15.1 (bug absent in 4.46.3)
- Python 3.12, Linux x86_64
- Model: `deepseek-ai/deepseek-math-7b-rl` (affects this llama-style ByteLevel-BPE tokenizer family)

### Reproduction

```python
from transformers import AutoTokenizer

tok = AutoTokenizer.from_pretrained("deepseek-ai/deepseek-math-7b-rl")
text = "we need to add 9 and 6"
ids = tok(text, add_special_tokens=False)["input_ids"]
print(repr(tok.decode(ids)))
print(tok.convert_ids_to_tokens(ids))
```

Output under 5.15.1:

```
'weneedtoadd9and6'
['w', 'ene', 'ed', 'to', 'add', '9', 'and', '6']
```

Under 4.46.3 the same code round-trips correctly (`'we need to add 9 and 6'`, with `Ġ`-prefixed word-boundary tokens).

### Expected behavior

Encoding should preserve whitespace (byte-level BPE with `Ġ` word-boundary markers). Under 5.15.1, `AutoTokenizer` appears to rebuild this tokenizer rather than using the repo's `tokenizer.json` verbatim, and the rebuilt version:

1. **strips all spaces during encoding** — the token sequence above corresponds to the space-stripped string, not the input; and
2. renders byte-level markers (`Ġ`, `Ċ`) **literally in decoded text** in generation pipelines.

This is silent (no warning), and generations from the garbled inputs remain semi-coherent, which makes it easy to miss — in our case it invalidated a full evaluation run before being caught.

### Workaround

Loading the repo's `tokenizer.json` verbatim gives correct behavior and is token-identical to 4.46.3's AutoTokenizer output (verified on ~4k prompts):

```python
from transformers import PreTrainedTokenizerFast
tok = PreTrainedTokenizerFast(tokenizer_file="<path>/tokenizer.json")
# graft chat_template / bos / eos from the (broken) AutoTokenizer load
```

Happy to provide more detail or test candidate fixes.

**Comment by Chirag6722** [comment]
Reproduced on `transformers` 5.15.1 / `tokenizers` 0.22.2, Python 3.12 (Windows). Your diagnosis is right, and the mechanism turns out to be specific enough to be worth writing down: the rebuilt tokenizer is a **sentencepiece/Metaspace** pipeline applied to a **ByteLevel-BPE** vocabulary.

## The two loaders disagree on the same file

```python
from tokenizers import Tokenizer
from huggingface_hub import hf_hub_download
from transformers import AutoTokenizer

text = "we need to add 9 and 6"
raw = Tokenizer.from_file(hf_hub_download("deepseek-ai/deepseek-math-7b-rl", "tokenizer.json"))
tok = AutoTokenizer.from_pretrained("deepseek-ai/deepseek-math-7b-rl")
```

```
raw tokenizer.json  -> ['we','Ġneed','Ġto','Ġadd','Ġ','9','Ġand','Ġ','6']   decode: 'we need to add 9 and 6'
AutoTokenizer       -> ['w','ene','ed','to','add','9','and','6']             decode: 'weneedtoadd9and6'
```

Same `tokenizer.json`, so nothing is wrong with the file.

## What gets replaced

Comparing the file against `tok._tokenizer.to_str()` after loading:

| component | in `tokenizer.json` | after `AutoTokenizer.from_pretrained` |
| --- | --- | --- |
| `pre_tokenizer` | `Sequence` of `Split` (ByteLevel BPE) | `Metaspace`, `replacement="▁"`, `prepend_scheme="always"`, `split=false` |
| `decoder` | `ByteLevel` | `Sequence[Replace("▁"→" "), ByteFallback, Fuse, …]` |
| `normalizer` | `Sequence` (empty) | `null` |

That is the Llama **sentencepiece** pipeline. The vocabulary it is applied to is byte-level, where word boundaries are `Ġ`, not `▁`. A `Metaspace` pre-tokenizer looking for `▁` finds none, so no token carries a boundary marker and the spaces are dropped — which is exactly the observed `['w','ene','ed',...]` segmentation of the space-stripped string. The `▁→" "` decoder rule then has nothing to substitute, so the round trip cannot recover them, and in generation the raw `Ġ`/`Ċ` markers surface literally.

Two details that may help locate it:

- **The repo has no `tokenizer.model`.** `hf_hub_download(..., "tokenizer.model")` 404s; only `tokenizer.json` (4.6 MB) is present. So a sentencepiece pipeline is being synthesised for a model that ships no sentencepiece source.
- **`tokenizer_config.json` says `"tokenizer_class": "LlamaTokenizerFast"`**, and the object returned reports `is_fast=True` with class `LlamaTokenizer` — so this is not a fast/slow fallback. The fast path itself is the one substituting the components.
- Passing `use_fast=True` and `use_fast=False` both produce the corrupted output, so there is no combination of that flag that avoids it.

## Why it is worth prioritising

As you note, it is silent. There is no warning, the output stays semi-coherent, and the failure is only visible if someone checks the round trip. Anyone fine-tuning or evaluating a `Ġ`-vocabulary Llama-style model on 5.x gets quietly wrong tokenisation, and the natural assumption is that the model is underperforming rather than that the input was mangled.

Your workaround (loading `tokenizer.json` verbatim via `PreTrainedTokenizerFast`) matches the raw output above, so it is a correct one in the meantime.

**Comment by HariomJangra** [comment]
> ### System Info
> * `transformers` 5.15.1 (bug absent in 4.46.3)
> * Python 3.12, Linux x86_64
> * Model: `deepseek-ai/deepseek-math-7b-rl` (affects this llama-style ByteLevel-BPE tokenizer family)
> 
> ### Reproduction
> from transformers import AutoTokenizer
> 
> tok = AutoTokenizer.from_pretrained("deepseek-ai/deepseek-math-7b-rl")
> text = "we need to add 9 and 6"
> ids = tok(text, add_special_tokens=False)["input_ids"]
> print(repr(tok.decode(ids)))
> print(tok.convert_ids_to_tokens(ids))
> Output under 5.15.1:
> 
> ```
> 'weneedtoadd9and6'
> ['w', 'ene', 'ed', 'to', 'add', '9', 'and', '6']
> ```
> 
> Under 4.46.3 the same code round-trips correctly (`'we need to add 9 and 6'`, with `Ġ`-prefixed word-boundary tokens).
> 
> ### Expected behavior
> Encoding should preserve whitespace (byte-level BPE with `Ġ` word-boundary markers). Under 5.15.1, `AutoTokenizer` appears to rebuild this tokenizer rather than using the repo's `tokenizer.json` verbatim, and the rebuilt version:
> 
> 1. **strips all spaces during encoding** — the token sequence above corresponds to the space-stripped string, not the input; and
> 2. renders byte-level markers (`Ġ`, `Ċ`) **literally in decoded text** in generation pipelines.
> 
> This is silent (no warning), and generations from the garbled inputs remain semi-coherent, which makes it easy to miss — in our case it invalidated a full evaluation run before being caught.
> 
> ### Workaround
> Loading the repo's `tokenizer.json` verbatim gives correct behavior and is token-identical to 4.46.3's AutoTokenizer output (verified on ~4k prompts):
> 
> from transformers import PreTrainedTokenizerFast
> tok = PreTrainedTokenizerFast(tokenizer_file="<path>/tokenizer.json")
> # graft chat_template / bos / eos from the (broken) AutoTokenizer load
> Happy to provide more detail or test candidate fixes.



The root issue is that `AutoTokenizer` / `TokenizersBackend` extracts post-processing and truncation settings when loading from `tokenizer.json`, but was omitting the saved `pre_tokenizer` and `decoder`. Because of this, `LlamaTokenizer` was forcing its default Metaspace (`▁`) pre-tokenizer and Replace decoder onto models that actually ship with a ByteLevel BPE pipeline (like `deepseek-math`).

The fix ensures `TokenizersBackend` preserves the `pre_tokenizer` and `decoder` from `tokenizer.json` when present, allowing `LlamaTokenizer` to respect them instead of overwriting them with Metaspace defaults.

I have verified the fix on `deepseek-ai/deepseek-math-7b-rl` and will open a PR shortly!

**Comment by gavinkvx** [comment]
Same root cause reproduces on `deepseek-ai/DeepSeek-Prover-V2-7B` (transformers 5.15.1, tokenizers 0.22.2, Python 3.12), which ships the same byte-level BPE `tokenizer.json` labelled `"tokenizer_class": "LlamaTokenizerFast"`.

```python
from transformers import AutoTokenizer
m = "deepseek-ai/DeepSeek-Prover-V2-7B"
s = "theorem t (a b : ℕ) : a + b = b + a := by\n  rw [add_comm]"
tok = AutoTokenizer.from_pretrained(m)
print(type(tok).__name__, tok._tokenizer.decoder)
# LlamaTokenizer Sequence(decoders=[Replace(pattern=String("▁"), content=" "), ByteFallback(), Fuse(), Strip(content=" ", start=1, stop=0)])
ids = tok.encode(s, add_special_tokens=False)
print(tok.convert_ids_to_tokens(ids)[:6])   # ['theorem', 't', '(', 'ab', ':', '):']
print(tok.decode(ids) == s)                 # False
```

So this is not only a decode problem: the SentencePiece-style pipeline installed by the `LlamaTokenizer` class replaces the ByteLevel pre-tokenizer/decoder defined in `tokenizer.json`, and whitespace plus non-ASCII (here `ℕ`) are already gone at encode time. Downstream, vLLM 0.27.1 serving this checkpoint emits raw `Ġ`/`Ċ` markers on `/v1/completions` and space-stripped text on `/v1/chat/completions`.

Workaround that restores a clean round-trip: copy the tokenizer files to a local directory and set `"tokenizer_class": "PreTrainedTokenizerFast"` in `tokenizer_config.json`. `AutoTokenizer` then resolves to `TokenizersBackend`, loads `tokenizer.json` as-is, and `decode(encode(s)) == s`. For vLLM, `--tokenizer <that directory>` is sufficient.

Separately, `print(tokenizers.Tokenizer.from_file("tokenizer.json").pre_tokenizer)` panics in `serde_pyo3.rs:153` (repr truncation at byte 100 lands inside a multi-byte char of the ByteLevel regex) — that is a `tokenizers` issue, mentioned only because anyone bisecting this will hit it.
